codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/Waiters.java

```diff
@@ -0,0 +1,835 @@
+                    );
+                    writer.write(
+                            "WaiterStateMutator func(context.Context, $P, $P, error) "
+                                    + "(bool, error)", inputSymbol, outputSymbol);
+                }
+        );
+        writer.write("");
+    }
+
+
+    /**
+     * Generates waiter client used to invoke waiter function. The waiter client is specific to a modeled waiter.
+     * Each waiter client is unique within a enclosure of a service.
+     * This function also generates a waiter client constructor that takes in a API client interface, and waiter options
+     * to configure a waiter client.
+     */
+    private void generateWaiterClient(
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            OperationShape operationShape,
+            String waiterName,
+            Waiter waiter
+    ) {
+        Symbol operationSymbol = symbolProvider.toSymbol(operationShape);
+        String clientName = generateWaiterClientName(waiterName);
+
+        writer.write("");
+        writer.writeDocs(
+                String.format("%s defines the waiters for %s", clientName, waiterName)
+        );
+        writer.openBlock("type $L struct {", "}",
+                clientName, () -> {
+                    writer.write("");
+                    writer.write("client $L", generateApiClientInterfaceName(operationSymbol));
+
+                    writer.write("");
+                    writer.write("options $L", generateWaiterOptionsName(waiterName));
+                });
+
+        writer.write("");
+
+        String constructorName = String.format("New%s", clientName);
+
+        Symbol waiterOptionsSymbol = SymbolUtils.createPointableSymbolBuilder(
+                generateWaiterOptionsName(waiterName)
+        ).build();
+
+        Symbol clientSymbol = SymbolUtils.createPointableSymbolBuilder(
+                clientName
+        ).build();
+
+        writer.writeDocs(
+                String.format("%s constructs a %s.", constructorName, clientName)
+        );
+        writer.openBlock("func $L(client $L, optFns ...func($P)) $P {", "}",
+                constructorName, generateApiClientInterfaceName(operationSymbol), waiterOptionsSymbol, clientSymbol,
+                () -> {
+                    writer.write("options := $T{}", waiterOptionsSymbol);
+                    writer.addUseImports(SmithyGoDependency.TIME);
+
+                    // set defaults
+                    writer.write("options.MinDelay = $L * time.Second", waiter.getMinDelay());
+                    writer.write("options.MaxDelay = $L * time.Second", waiter.getMaxDelay());
+                    // set defaults not defined in model
+                    writer.write("options.MaxWaitTime = $L * time.Second", DEFAULT_MAX_WAIT_TIME_IN_SECONDS);
+                    writer.write("options.MaxAttempts = $L ", DEFAULT_MAX_ATTEMPTS);
```

Comment:
Does smithy not model `MaxAttempts`? I believe this was a value that was modeled by the v1 SDK's waiter definitions.

https://github.com/aws/aws-sdk-go/blob/98d117b72099e244e566a83870e141cde6f7d8a7/models/apis/s3/2006-03-01/waiters-2.json#L4-L8